rust/src/mask/model.rs

```diff
@@ -0,0 +1,377 @@
+        Box::new(self.0.into_iter().map(|i| {
+            i.to_integer().to_i32().ok_or_else(|| ModelCastError {
+                weight: i,
+                target: PrimitiveType::I32,
+            })
+        }))
+    }
+}
+
+impl FromPrimitives<i32> for Model {
+    fn from_primitives<I: Iterator<Item = i32>>(iter: I) -> Result<Self, PrimitiveCastError<i32>> {
+        Ok(iter.map(|p| Ratio::from_integer(BigInt::from(p))).collect())
+    }
+
+    fn from_primitives_bounded<I: Iterator<Item = i32>>(iter: I) -> Self {
+        Self::from_primitives(iter).unwrap()
+    }
+}
+
+impl IntoPrimitives<i64> for Model {
+    fn into_primitives(self) -> Box<dyn Iterator<Item = Result<i64, ModelCastError>>> {
+        Box::new(self.0.into_iter().map(|i| {
+            i.to_integer().to_i64().ok_or_else(|| ModelCastError {
+                weight: i,
+                target: PrimitiveType::I64,
+            })
+        }))
+    }
+}
+
+impl FromPrimitives<i64> for Model {
+    fn from_primitives<I: Iterator<Item = i64>>(iter: I) -> Result<Self, PrimitiveCastError<i64>> {
+        Ok(iter.map(|p| Ratio::from_integer(BigInt::from(p))).collect())
+    }
+
+    fn from_primitives_bounded<I: Iterator<Item = i64>>(iter: I) -> Self {
+        Self::from_primitives(iter).unwrap()
+    }
+}
+
+impl IntoPrimitives<f32> for Model {
+    fn into_primitives(self) -> Box<dyn Iterator<Item = Result<f32, ModelCastError>>> {
+        let iter = self.0.into_iter().map(|r| {
+            ratio_to_float::<f32>(&r).ok_or_else(|| ModelCastError {
+                weight: r,
+                target: PrimitiveType::F32,
+            })
+        });
+        Box::new(iter)
+    }
+}
+
+impl FromPrimitives<f32> for Model {
+    fn from_primitives<I: Iterator<Item = f32>>(iter: I) -> Result<Self, PrimitiveCastError<f32>> {
+        iter.map(|f| float_to_ratio::<f32>(f).ok_or_else(|| PrimitiveCastError(f)))
+            .collect()
+    }
+
+    fn from_primitives_bounded<I: Iterator<Item = f32>>(iter: I) -> Self {
+        iter.map(float_to_ratio_bounded::<f32>).collect()
+    }
+}
+
+impl IntoPrimitives<f64> for Model {
+    fn into_primitives(self) -> Box<dyn Iterator<Item = Result<f64, ModelCastError>>> {
+        let iter = self.0.into_iter().map(|r| {
+            ratio_to_float::<f64>(&r).ok_or_else(|| ModelCastError {
+                weight: r,
+                target: PrimitiveType::F64,
+            })
+        });
+        Box::new(iter)
+    }
+}
+
+impl FromPrimitives<f64> for Model {
+    fn from_primitives<I: Iterator<Item = f64>>(iter: I) -> Result<Self, PrimitiveCastError<f64>> {
+        iter.map(|f| float_to_ratio::<f64>(f).ok_or_else(|| PrimitiveCastError(f)))
+            .collect()
+    }
+
+    fn from_primitives_bounded<I: Iterator<Item = f64>>(iter: I) -> Self {
+        iter.map(float_to_ratio_bounded::<f64>).collect()
+    }
+}
+
+fn ratio_to_float<F: FloatCore>(ratio: &Ratio<BigInt>) -> Option<F> {
```

Comment:
That was what I did initially but ran into an issue, I can't remember what. I'll try it and report back.

Reply:
Yeah so in the iterator I actually want to move the ratio in case conversion fails. That means I should only give a reference to `ratio_to_float`:

```
        let iter = self.0.into_iter().map(|r| {
                                  +---- I need to borrow here
                                  v
            ratio_to_float::<f32>(&r).ok_or_else(|| ModelCastError {
                weight: r, // because this is a move
                target: PrimitiveType::F32,
            })
        });
```